Enqueue to validation queue when a package is created

joelverhagen opened the pull request:
Depends on PR https://github.com/NuGet/NuGetGallery/pull/4665.

Progress on https://github.com/NuGet/NuGetGallery/issues/4646.

joelverhagen commented on src/NuGetGallery/Services/AsynchronousPackageValidationInitiator.cs:
```diff
@@ -0,0 +1,30 @@
+
+        public async Task StartValidationAsync(Package package)
+        {
+            package.PackageStatusKey = PackageStatus.Available;
```
For now, mark packages as `Available` immediately. Only during the final phase of rolling this feature out will we set the package status to `Validating` here.

chenriksson replied:
Available is the default, so this shouldn't be necessary - right? Makes me wonder if we should have a different, disabled default that requires us to explicitly set this to Available.

joelverhagen replied:
`Available` is default, yes. I think this should remain the default since it matches the DB default. I put this line here more-so to be defensive and to call out the location where we will soon have `... = PackageStatus.Validating`.

joelverhagen commented on src/NuGetGallery/packages.config:
```diff
@@ -82,10 +82,13 @@
   <package id="NuGet.Services.Logging" version="2.2.3.0" targetFramework="net46" />
+  <package id="NuGet.Services.Owin" version="2.2.3" targetFramework="net46" />
   <package id="NuGet.Services.Platform.Client" version="3.0.29-r-master" targetFramework="net46" />
+  <package id="NuGet.Services.ServiceBus" version="2.2.3-dev-13985" targetFramework="net46" />
```
Before merging, I will publish a stable version of these new NuGet.Service.* dependencies.

skofman1 commented on src/NuGetGallery/Configuration/AppConfiguration.cs:
```diff
@@ -66,6 +66,14 @@ public class AppConfiguration : IAppConfiguration
+        public bool AsynchronousPackageValidationEnabled { get; set; }
+
+        [DisplayName("AzureServiceBus.Validation.ConnectionString")]
+        public string AzureServiceBus_Validation_ConnectionString { get; set; }
```
Can we introduce a class to contain all service bus config?

joelverhagen replied:
I would rather not. This is a case very similar to our Azure Storage configuration, which is also in `IAppConfiguration`. Until we have formal design for re-organization of configuration I prefer to align with what we already have for Azure Storage. Otherwise, we may end up having multiple approaches done here.

joelverhagen replied:
I'll see if I can come up with a good place to put these configuration values that can easily be adapted by other groupings.

joelverhagen replied:
I've added `IServiceBusConfiguration`.

skofman1 commented on src/NuGetGallery/App_Start/DefaultDependenciesModule.cs:
```diff
@@ -314,6 +318,38 @@ protected override void Load(ContainerBuilder builder)
+        private void RegisterAsynchronousValidation(ContainerBuilder builder, ConfigurationService configuration)
+        {
+            builder
+                .RegisterType<ServiceBusMessageSerializer>()
```
Can we avoid making the code aware that we are using service bus? Won't IValidationMessageSerializer do the job?

joelverhagen replied:
DI container code needs to be aware, unless we want to implement modules in the dependent libraries, which is worse since that would mean Autofac dependency goes everywhere.

Concerning the name, the serialization of messages has close affinity with the queue implementation. For example, Service Bus SDK uses data contract serialization for data and has a property bag for metadata. Other queue implement X may require entirely different serialization methods.

Note that taking a data model `PackageValidationMessageData` and tossing it on the queue is abstracted away at the `IPackageValidationEnqueuer` level. The implementation of `IPackageValidationInitiator` does not know anything about Service Bus.

There is a middle ground where we try to make a good guess on how to serialize messages by, say turning each data object into a `string` of data and `Dictionary<string, string>` for metadata, but this is making assumptions we have no reason to make. Even if we do this, we would still need an implementation to turn these two things (`string` and `Dictionary<string, string>`) to a Service Bus `BrokeredMessage`. This would still be visible here in the DI set up and would very likely have the same name.

skofman1 replied:
Spoke offline.

shishirx34 commented on src/NuGetGallery.Core/PackageStatus.cs:
```diff
@@ -15,5 +15,19 @@ public enum PackageStatus
+        /// <summary>
+        /// The package is in the process of being validated. The validation is being done asynchronously. When the
+        /// validation is complete it will either move to the <see cref="Available"/> or the <see cref="Deleted"/>
+        /// state. When packages are in this state, they are not available but the package ID and version are still
```
Can't the state move to `FailedValidation` state as well in here?

joelverhagen replied:
Yes, nice catch.

chenriksson commented on src/NuGetGallery/App_Start/DefaultDependenciesModule.cs:
```diff
@@ -314,6 +318,38 @@ protected override void Load(ContainerBuilder builder)
+                .As<IServiceBusMessageSerializer>();
+
+            builder
+                .RegisterType<PackageValidationEnqueuer>()
```
Are the serializer and enqueuer part of the NuGet.Services.* dependencies? Assuming these are in separate PR / repo?

joelverhagen replied:
They are in the ServerCommon repo:
https://github.com/NuGet/ServerCommon/tree/dev/src

joelverhagen replied:
If the shape of the interfaces seems incorrect, please let me know on this PR. I will apply the changes in all of the necessary places.

cristinamanum commented on src/NuGetGallery/Services/PackageService.cs:
```diff
@@ -138,6 +111,8 @@ public void EnsureValid(PackageArchiveReader packageArchiveReader)
             packageRegistration.Packages.Add(package);
             await UpdateIsLatestAsync(packageRegistration, commitChanges: false);
 
+            await _validationInitiator.StartValidationAsync(package);
```
Will it be possible for this await to throw an exception? If yes what is the status of the package on the failure? Will it be actionable for the user?

joelverhagen replied:
If we fail to write to Service Bus an exception will be thrown. The package has not been written to DB or packages container yet so there is no concern of inconsistency.

scottbommarito commented on src/NuGetGallery/App_Start/DefaultDependenciesModule.cs:
```diff
@@ -314,6 +318,38 @@ protected override void Load(ContainerBuilder builder)
+        private void RegisterAsynchronousValidation(ContainerBuilder builder, ConfigurationService configuration)
+        {
+            builder
+                .RegisterType<ServiceBusMessageSerializer>()
```
Can we move the registration of `ServiceBusMessageSerializer` and `PackageValidationEnqueuer` inside the if statement (`AsynchronousPackageValidationEnabled`)? Are either of these needed if `AsynchronousPackageValidationEnabled` is disabled?

joelverhagen replied:
I considered that but I thought it would be easier to put as much as possible into the DI container and let the run time figure out what is necessary. That's easier than trying to think through the minimal set of types required for activation. Only in cases where we choose between implementation A or B should we consider not adding something to the container.

scottbommarito commented on src/NuGetGallery/Web.config:
```diff
@@ -543,6 +547,30 @@
   <system.serviceModel>
     <serviceHostingEnvironment aspNetCompatibilityEnabled="true" multipleSiteBindingsEnabled="true" />
+    <extensions>
+      <!-- In this extension section we are introducing all known service bus extensions. User can remove the ones they don't need. -->
```
Can you explain exactly what these are for? Is this implying that we need to add these extensions to use service bus?

joelverhagen replied:
These are added to the web.config by the Service Bus package. I would rather leave the defaults as-is. From what I can tell from online documentation, they basically enable features that come along with `Microsoft.ServiceBus`.

scottbommarito replied:
Does this comment come automatically with the service bus package?

Because I'd prefer to see something like
```
<!-- Service Bus extensions (added automatically by the Azure.ServiceBus package) -->
```

joelverhagen replied:
> Does this comment come automatically with the service bus package?

Yes.

> Because I'd prefer to see something like

Sure.

scottbommarito commented on tests/NuGetGallery.Facts/Services/AsynchronousPackageValidationInitiatorFacts.cs:
```diff
@@ -0,0 +1,96 @@
+            }
+        }
+
+        public abstract class Base
```
I was confused at first where these fields were coming from.

I suggest
1 - Changing `Base` to something more descriptive (e.g. `MethodFactsBase`).
2 - Moving it to the top of the file so it the first thing people read when try to read this file.

Same for `ImmediatePackageValidatorFacts.Base` as well.

joelverhagen replied:
I think it's more confusing to put test infrastructure above the tests. Facts/asserts/etc are the most important part of a test class and should therefore be most visible.

chenriksson commented on src/NuGetGallery/Web.config:
```diff
@@ -38,6 +38,10 @@
+
+    <add key="Gallery.AsynchronousPackageValidationEnabled" value="false" />
+    <add key="AzureServiceBus.Validation.ConnectionString" value="" />
+    <add key="AzureServiceBus.Validation.TopicName" value="" />
```
Do we need the async flag, or can we assume that if the validation connection string is provided?

chenriksson replied:
If ConnectionString is missing, I'm assuming package pushes will fail and we'll see exceptions in AI.

If TopicName is missing, would consumers just not receive the messages? Should we verify that this config is set if async is enabled?

joelverhagen replied:
I'll leave some comments about this. Essentially, if `Gallery.AsynchronousPackageValidationEnabled` is `true` both the `ConnectionString` and `TopicName` must be set.

joelverhagen commented on src/NuGetGallery/Services/AsynchronousPackageValidationInitiator.cs:
```diff
@@ -0,0 +1,30 @@
+                package.PackageRegistration.Id,
+                package.Version,
+                Guid.NewGuid());
+            await _enqueuer.StartValidationAsync(data);
```
Yes, great idea!

skofman1 commented on src/NuGetGallery/Configuration/ConfigurationService.cs:
```diff
@@ -21,29 +21,27 @@ public class ConfigurationService : PoliteCaptcha.IConfigurationSource, IGallery
+        private readonly Lazy<ISecretInjector> _secretInjector;
+        private readonly Lazy<IAppConfiguration> _lazyAppConfiguration;
+        private readonly Lazy<FeatureConfiguration> _lazyFeatureConfiguration;
+        private readonly Lazy<IServiceBusConfiguration> _lazyServiceBusConfiguration;
```
>IServiceBusConfiguration [](start = 30, length = 24)

Like :)

skofman1 commented on src/NuGetGallery/Configuration/IServiceBusConfiguration.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public interface IServiceBusConfiguration
+    {
+        string Validation_ConnectionString { get; set; }
```
Why do we need the validation prefix?

joelverhagen replied:
When we have other Service Bus integrations, this will be useful to disambiguate: for example, when we disconnect V3 from V2 Feed.

skofman1 replied:
If the plan is to have a separate interface and implementation pair for each service bus connection, it seems redundant. You can simply have a class called ValidationServiceBusConfiguration.
On the other hand, if we want all service bus config classes to have a shared interface, we can achieve that as well using DI.

joelverhagen replied:
I think we should group all Service Bus configuration keys in the `IServiceBusConfiguration` interface. I don't see much value in a one or two property configuration interface.

agr commented on src/NuGetGallery/Services/IPackageValidationInitiator.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace NuGetGallery
+{
+    public interface IPackageValidationInitiator
```
Documentation?..

joelverhagen replied:
Done.

joelverhagen commented:
I have made a change where we will not enqueue if the gallery is in read-only mode. This saves us from enqueuing a non-actionable Service Bus message on nuget.exe push against a read-only gallery, since the DB commit will fail.

skofman1 commented on src/NuGetGallery/Services/AsynchronousPackageValidationInitiator.cs:
```diff
@@ -15,18 +16,26 @@ namespace NuGetGallery
         {
+            if (_appConfiguration.ReadOnlyMode)
+            {
+                throw new ReadOnlyModeException("Enqueuing unavailable: the gallery is currently in read only mode, with limited service. Please try again later.");
```
Can we move this message to Strings?

joelverhagen replied:
Sure, but note that this string is not user-visible. It will only appear in our exception logs. I think it makes sense for it to be in `Strings` since other enqueuers might use it.

joelverhagen replied:
Fixed.

joelverhagen commented:
This merge is pending a stable version of `NuGet.Services.Validations`.

joelverhagen commented:
Everyone, I have made some significant changes since I first sent out this PR. 

- Removed `commitChanges` parameter from some places where it was always `false`. This was necessary so that `PackageUploadService` could handle the Service Bus enqueueing for validation instead of `PackageService` (PackageService.CreatePackageAsync(..., commitChanges: true) was preventing this move).
- Introduced `IValidationService` for all the validation-y goodness. This will have admin and revalidation logic.

loic-sharma commented on tests/NuGetGallery.Facts/Services/ValidationServiceFacts.cs:
```diff
@@ -0,0 +1,68 @@
+
+                // Assert
+                _packageService.Verify(
+                    x => x.UpdatePackageStatusAsync(_package, packageStatus, false),
```
>false [](start = 77, length = 5)

Please add a named parameter for the `false` :)

joelverhagen replied:
Doesn't work in `Expression`s

loic-sharma replied:
Huh, I didn't know that. Good to know.

cristinamanum commented on src/NuGetGallery/Services/IValidationService.cs:
```diff
@@ -0,0 +1,20 @@
+        /// commit to the database is the responsibility of the caller.
+        /// </summary>
+        /// <param name="package">The package to start validation for.</param>
+        Task StartValidationForNewPackageAsync(Package package);
```
Why the name StartValiadtionForNewPackage and not only StartValidationForPackage?

joelverhagen replied:
I've changed the method name to `StartValidationAsync`.

skofman1 commented on src/NuGetGallery/Controllers/ApiController.cs:
```diff
@@ -482,8 +482,7 @@ public virtual Task<ActionResult> CreatePackagePost()
                             packageStreamMetadata,
-                            user,
-                            commitChanges: false);
+                            user);
```
Although commitChanges is always false, I think it adds to the readability of the code. Consider adding it back.

joelverhagen replied:
Dead code should not be maintained! We should get into the mindset of having limited locations where we commit changes.

joelverhagen commented on src/NuGetGallery/Services/ReservedNamespaceService.cs:
```diff
@@ -228,10 +228,8 @@ public class ReservedNamespaceService : IReservedNamespaceService
         /// </summary>
```
Done.

joelverhagen commented:
Merged with an unsigned version of ServerCommon packages. This matches existing behavior with `NuGet.Services.KeyVault`. This feature is currently disabled in DEV, INT, and PROD.
